### ISSUE TITLE
Dynamic compressor supports leads to error during loading certificate

### ISSUE THREAD

**Issue by sebastianas** [openssl/openssl#23558]
I configured openssl 3.2.1 with enable-brotli-dynamic. In the resulting filesystem is libbrotli not installed. This leads to a failure in dovecot which can not start a server on a TLS port. The reason is that dovecot checks openssl error code via ERR_peek_error() after invoking SSL_CTX_use_certificate(). See
   https://github.com/dovecot/core/blob/fe086cfa085fc5190f233993cb523a8678014681/src/lib-ssl-iostream/iostream-openssl-context.c#L224

SSL_CTX_use_certificate() tried to load brotli support, failed, set error but returned without an error as return code (the function was successful as per return code). Since ERR_peek_error() returns an error, dovecot complains and reports:

>  dovecot: imap-login: Error: Failed to initialize SSL server context:
>  Can't load SSL certificate (ssl_cert setting): error:12800067:DSO
>  support routines::could not load the shared library:
>  filename(libbrotlienc.so): libbrotlienc.so: cannot open shared object
>  file: No such file or directory, error:12800067:DSO support
>  routines::could not load the shared library, error:12800067:DSO support
>  routines::could not load the shared library: filename(libbrotlidec.so):
>  libbrotlidec.so: cannot open shared object file: No such file or
>  directory, error:12800067:DSO support routines::could not load the
>  shared library: user=<>, rip=127.0.0.1, lip=127.0.0.1,
>  session=<uoxhJyERNuesewoJ>

Removing the ERR_peek_error() part in dovecot fixes the problem.
The dovecot function is based on use_certificate_chain_file() from OpenSSL which as the same check:
  https://github.com/openssl/openssl/blob/7b4fc8eb72cf5b2de7f2b22838449d28cfd9f0c7/ssl/ssl_rsa.c#L492

Indeed OpenSSL testsuite fails for instance in 80-test_ssl_old.t after removing libbrotli:

```
     # Testing ciphersuites for -tls1_3 
 Warning: unknown compression not supported
 # 403791168B7F0000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:../crypto/dso/dso_dlfcn.c:118:filename(libbr otlidec.so): libbrotlidec.so: cannot open shared object file: No such file or directory
 # 403791168B7F0000:error:12800067:DSO support routines:DSO_load:could not load the shared library:../crypto/dso/dso_lib.c:147:
 # 403791168B7F0000:error:0A0000B9:SSL routines:SSL_CTX_set_cipher_list:no cipher match:../ssl/ssl_lib.c:3358:
 ../../util/wrap.pl ../../test/ssl_old_test -s_key ../../../test/certs/ee-key.pem -s_cert certU.ss -c_key ../../../test/certs/ee-key.pem -c_cert  certU.ss -config ../../../test/default-and-legacy.cnf -provider default -provider legacy -s_cert certD.ss -s_key keyD.ss -s_cert certE.ss -s_key  keyE.ss -cipher '' -ciphersuites TLS_AES_256_GCM_SHA384 => 1
     not ok 119 - Testing 
     
     #   Failed test 'Testing '
     #   at ../test/recipes/80-test_ssl_old.t line 507.
```

How do we deal with this? Do we avoid setting an error if using a dynamic compressor is not mandatory?

Sebastian

**Comment by t8m** [comment]
Yeah, IMO when loading the compressing library dynamically on demand we should not set an error. (Or rather by using ERR_set_mark() ERR_pop_to_mark() we should remove the errors if we do not return a failure from the function.)

**Comment by sebastianas** [comment]
If this is what we want then I would do something similar for zlib/ zstd so that it behaves the same way.

**Comment by fleximus** [comment]
Still present with OpenSSL 3.3.2 and `libzstd.so`
> imap-login: Error: Failed to initialize SSL server context: Can't load SSL certificate (ssl_cert setting): error:12800067:DSO support routines::could not load the shared library: filename(libzstd.so): Shared object "libzstd.so" not found, required by "imap-login", error:12800067:DSO support routines::could not load the shared library: user=<>, rip=127.0.0.1, lip=127.0.0.1, session=

**Comment by paulzzh** [comment]
Any progress here? I found [this](https://salsa.debian.org/debian/openssl/-/commit/26109eddd699cce5c3aecb42ec1c5178cbde7d84) while trying to enable nginx’s [certificate compression](https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_certificate_compression) support for [Chromium](https://blog.nginx.org/blog/nginx-open-source-1-29-3-and-1-29-4).

> The directive is supported when using OpenSSL 3.2 or higher; the list of supported compression algorithms is **provided by the library**.

> Chrome and Chromium-based browsers only advertise support for Brotli.
